src/mame/sharp/x68k_v.cpp

```diff
@@ -426,7 +426,7 @@ bool x68k_state::draw_gfx(bitmap_rgb32 &bitmap,rectangle cliprect)
 	return gfxblend;
 }
 
-template <bool Blend> rgb_t x68k_state::get_gfx_pixel(int scanline, int pixel, bool gfxblend, rgb_t blendpix)
+rgb_t x68k_state::get_gfx_pixel(int scanline, int pixel, bool gfxblend, rgb_t blendpix, bool textblend)
```

Comment:
The point of using a template for this is that the compiler will instantiate versions of it for both cases, and conditional branches inside the function will be avoided.  You’re still only invoking it from places where the value of `textblend` is already known from a prior check, so additional checks inside the function are unnecessary.  Also, you’re now only calling it when `textblend` is true, so the compiler won’t even need to instantiate ` template <bool Blend> rgb_t x68k_state::get_gfx_pixel`.  There’s no benefit to changing from a template parameter to a function parameter.

Reply:
I'm happy to revert this if there's a preference for style, but in my testing there's a performance increase with a single function and textblend passed as an argument (on my PC, 434% vs. 410% on an Akumajou Dracula benchmark). Whether that comes from the ability of the compiler to optimise differently vs. runtime branch prediction, I'm not sure.

Style was the main reason I had a preference for the change though - I haven't seen templates used to pass fixed values to functions before, I'm used to templating based on types, and a (perhaps const) function argument is easier for me to understand here.

As an argument for the change, when textblend is false, blendpix is never used, and in the templated version of the function would never even be referenced.

The template was introduced by @cracyc in https://github.com/mamedev/mame/commit/1163278a7b74fced26f7749ebbeb406a65b481f9, so if they have any reasoning behind it that may help.

If you still have a preference for the template version with that explanation, let me know and I'm happy to change it back. If it's okay to leave in, I'll update the commit text.

> you’re now only calling it when textblend is true

Line 798 pulls the gfx screen's pixel and uses it as the final output value if the loop is at the gfx priority and blending gfx against text/pcg is disabled. I passed the variable by value after it had been assessed to be false rather than just using "false" as a reminder to what the function signature is (other style guides I've seen would use `/*textblend=*/false`), if that's confusing I can change it to a literal.

Reply:
> Style was the main reason I had a preference for the change though - I haven't seen templates used to pass fixed values to functions before, I'm used to templating based on types, and a (perhaps const) function argument is easier for me to understand here.

We template performance-critical functions on values that are known at the call site as it provides a performance improvement by avoiding redundant conditional branches.  Examples include:
* The [DSP16 CPU core](https://github.com/mamedev/mame/blob/master/src/devices/cpu/dsp16/dsp16.cpp) has the run loop templated as it’s very hot for a high-speed DSP.
* The [Hyperstone CPU core](https://github.com/mamedev/mame/blob/master/src/devices/cpu/e132xs/e132xsop.hxx) has the instruction handlers templated on the register banks, as this will already be known when the handler is dispatched.

Both of these have been extensively profiled and shown to perform better if the functions can be specialised at compile time.

Conditional branches, like anything that’s difficult for the instruction scheduler, incur a performance penalty on modern CPUs.  Things like the [delegate class](https://github.com/mamedev/mame/blob/master/src/lib/util/delegate.h) go out of their way to keep conditional branches off the hot path, in that case largely to mitigate the well-known performance penalty of the conditional branch when calling a pointer-to-member function with the Itanium C++ ABI.  In general, templating functions on values that will be known at the call site improves performance, unless you end up increasing the size of the generated code to the point that you exacerbate cache instruction cache misses (or worse, TLB misses).

> The template was introduced by @cracyc in 1163278a7b74fced26f7749ebbeb406a65b481f9, so if they have any reasoning behind it that may help.

I assume it was for performance reasons, to get the redundant checks out of  `get_gfx_pixel`, because the value is always known at the call site.

On further examination, it doesn’t actually save conditional branches in this case.  The two expressions involving `Blend` have no side effects, so it can evaluate both sides unconditionally and use a conditional move to set the local `blend`.  There’s still a single unavoidable conditional branch based on the local `blend` on either path.  So the only thing it’s saving is one stack-based parameter and a couple of logical instructions.

> Line 798 pulls the gfx screen's pixel and uses it as the final output value if the loop is at the gfx priority and blending gfx against text/pcg is disabled.

Sorry, I missed that.  I see now it’s being called once each for `true` and `false`.

I still find it very difficult to believe that it loses 5% performance by having two specialisations of this function.  How big is the generated code for that function?  Is it really causing enough icache evictions to have that much impact?  Are you accounting for thermal effects, clock boost, etc.?

Reply:
Just to clarify, “evaluate both sides unconditionally” refers to the expressions on both sides of the `&&` operator.